tox_core/src/dht/forced_ktree.rs

```diff
@@ -0,0 +1,438 @@
+
+    * its [`kbucket index`](./fn.kbucket_index.html) is lower than the
+      number of buckets.
+    * [`Bucket`](./struct.Bucket.html) to which it is added has free space
+      or added node is closer to the PK than other node in the bucket.
+
+    Returns `true` if node was added successfully, `false` otherwise.
+    */
+    pub fn try_add(&mut self, node: PackedNode) -> bool {
+        if self.ktree.try_add(node) {
+            if let Some(dht_node) = self.kbucket.remove(&self.ktree.pk, &node.pk) {
+                let added_node = self.ktree.get_node_mut(&node.pk).expect("Node should be added");
+                if node.saddr.is_ipv4() {
+                    added_node.assoc6 = dht_node.assoc6;
+                } else {
+                    added_node.assoc4 = dht_node.assoc4;
+                }
+            }
+            true
+        } else if !self.ktree.contains(&node.pk) {
+            self.kbucket.try_add(&self.ktree.pk, node, /* evict */ true)
+        } else {
+            false
+        }
+    }
+
+    /// Remove [`DhtNode`](./struct.DhtNode.html) with given PK from the
+    /// `Kbucket`.
+    pub fn remove(&mut self, node_pk: &PublicKey) -> Option<DhtNode> {
+        self.ktree.remove(node_pk).or_else(||
+            self.kbucket.remove(&self.ktree.pk, node_pk)
+        )
+    }
+
+    /** Get (up to) `count` closest nodes to given PK.
+
+    Functionality for [`SendNodes`](./struct.SendNodes.html).
+
+    Returns less than `count` nodes only if `Ktree` contains less than `count`
+    nodes.
+
+    It should not contain LAN ip node if the request is from global ip.
+    */
+    pub fn get_closest(&self, pk: &PublicKey, count: u8, only_global: bool) -> Kbucket<PackedNode> {
+        let mut kbucket = self.ktree.get_closest(pk, count, only_global);
+        for node in self.kbucket.iter().filter(|node| !node.is_bad()) {
+            if let Some(pn) = node.to_packed_node() {
+                if !only_global || IsGlobal::is_global(&pn.saddr.ip()) {
+                    kbucket.try_add(pk, pn, /* evict */ true);
+                }
+            }
+        }
+        kbucket
+    }
+
+    /**
+    Check if `Kbucket` contains [`PackedNode`] with given PK.
+
+    [`PackedNode`]: ./struct.PackedNode.html
+    */
+    pub fn contains(&self, pk: &PublicKey) -> bool {
+        self.kbucket.contains(&self.ktree.pk, pk) ||
+            self.ktree.contains(pk)
+    }
+
+    /**
+    Naive check whether a [`PackedNode`] can be added to the `Kbucket`.
+
+    Returns `true` if [`Bucket`] where node could be placed is not full
+    and node is not already in the [`Bucket`].
+
+    Otherwise `false` is returned.
+
+    [`Bucket`]: ./struct.Bucket.html
+    [`PackedNode`]: ./struct.PackedNode.html
+    */
+    pub fn can_add(&self, new_node: &PackedNode) -> bool {
+        self.ktree.can_add(new_node) ||
+            !self.ktree.contains(&new_node.pk) && self.kbucket.can_add(&self.ktree.pk, new_node, /* evict */ true)
```

Comment:
Is this is a correct check?

Reply:
Yes, we don't want to add node twice, so either it can be added to internal ktree (and removed from self if it's there) or to self.

Reply:
I mean can't we simplify to just "can_add || can_add"?

Reply:
We can't because if ktree can't add a node it can either be full or contain the node. If it's just full we might add the node to self kbucket, but if it already contains the node we don't won't to add it second time.

Reply:
That's right but kbucket has `, /* evict */ true` param whereas ktree hasn't. I wonder whether it worth checking that kbucket does not contain the node.

Reply:
We check that ktree doesn't contain it. It worth, otherwise it can be added twice.

Reply:
However with `self.kbucket.can_add(&self.ktree.pk, new_node, /* evict */ true)` it can be added twice

Reply:
No, it can't. It will overwrite previous added node. But if we don't check it it will be literally added twice, so we will send same packets to the same node twice.